Rewrote postinstall to Python (3.6+)

ocket8888 opened the pull request:
## What does this PR (Pull Request) do?
This PR features a rewrite of postinstall to Python3 (specifically the `_postinstall` Perl script). This is important because the Perl version of the script depended upon multiple libraries installed as dependencies of the Perl version of Traffic Ops - and if we are ever to get rid of that TO version things that depend on its dependencies will need to be reworked.

The new `postinstall.py` script has no dependencies outside of the Python3 standard library, and features several unit tests and an end-to-end test bash script whereas the Perl version had only a single unit test (executed at run time on every run).

To be clear, though, it - like the Perl script it is intended to replace - has dependencies on external programs including `wget`, `openssl` and `psql`. `wget` is only needed for Maxmind download support, so it could possibly be removed as the hard-coded URLs don't work anyway, but is left in for consistency with the old script. Besides that, its only external dependencies are already dependencies of TO - even the Go version - and the Python3 interpreter itself.

Finally, I should mention that this doesn't actually replace the old script, which is totally untouched, but provides the new Python script as an alternative. Many machines running TO in the wild may not have Python3 installed yet, so it may be too soon yet to require that for TO installations/upgrades.

- [x] This PR is not related to any Issue

## Which Traffic Control components are affected by this PR?
- Traffic Ops

## What is the best way to verify this PR?
To run the doctests, use `python -m doctest path/to/postinstall.py`. That should output nothing and exit successfully (my `python` is a Python3 interpreter, if yours is Python2 then ~~upgrade please, Python2 is dead~~ you'll need to use `python3 -m doctest path/to/postinstall.py` instead).

To run the end-to-end tests, use the `postinstall.test.sh` script, which should also output nothing (except spitting out some ssl keys/certs/requests in the current directory) and exit successfully.

## The following criteria are ALL met by this PR
- [x] This PR includes tests
- [x] Documentation is unnecessary - this isn't actually used yet. It does have docstrings and a `-h`/`--help` option, though
- [x] An update to CHANGELOG.md is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

zrhoffman commented on traffic_ops/install/bin/postinstall.py:
```diff
@@ -0,0 +1,1248 @@
+#!/usr/bin/python3
```
If we want to support python2, too, this should be just `python`, not `python3`.

CentOS 7 and CentOS 8 both have a python interpreter at `/usr/libexec/platform-python` (on CentOS 7 it is Python 2.7, on CentOS 8, it is Python 3.6). To accept a binary named `python` if it exists but falls back to `platform-python`, the shebang could be something like

```shell
#!/usr/bin/env -S sh -c 'PATH="${PATH}:/usr/libexec" exec "$(type -p python platform-python | head -n1)" "$0" "$@"'
```

zrhoffman reviewed (changes requested):
Python 2.7 is supported by Red Hat [through June 2024](https://access.redhat.com/support/policy/updates/rhel8-app-streams-life-cycle#th-collect19-b). Although we definitely don't *need* to support Python 2.7 just for this script to exist in the repo, there would be a much stronger case for this Python script to replace the Perl one if it could be run using both Python 2.7 and 3.6.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+
+"$MY_DIR/postinstall.py" --no-root --root-directory="$ROOT_DIR" --no-restart-to --no-database --ops-user="$(whoami)" --ops-group="$(whoami)" --automatic --cfile="$ROOT_DIR/defaults.json" --debug 2>"$ROOT_DIR/stderr" | tee "$ROOT_DIR/stdout"
+
+if [[ ! -z "$(grep 'ERROR' $ROOT_DIR/stderr)" ]]; then
```
No need for a string comparison, this can just be `if grep -q [...]`

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+mkdir -p "$ROOT_DIR/opt/traffic_ops/app/public/routing";
+mkdir "$ROOT_DIR/opt/traffic_ops/app/db";
+mkdir -p "$ROOT_DIR/opt/traffic_ops/app/conf/production";
+echo "{\"hypnotoad\":{\"listen\":[\"https://[::]:60443?cert=$ROOT_DIR/etc/pki/tls/certs/localhost.crt&key=$ROOT_DIR/etc/pki/tls/private/localhost.key\"]}}" > "$ROOT_DIR/opt/traffic_ops/app/conf/cdn.conf";
```
This would do better as `cat` with a here document so the double quotes don't need to be backslashed (and, optionally, so the JSON can be formatted)

ocket8888 replied:
Doing that causes the test to fail to find `users.json`:
```diff
diff --git a/traffic_ops/install/bin/postinstall.test.sh b/traffic_ops/install/bin/postinstall.test.sh
index d98c11645..cb8405c12 100755
--- a/traffic_ops/install/bin/postinstall.test.sh
+++ b/traffic_ops/install/bin/postinstall.test.sh
@@ -32,7 +32,15 @@ mkdir "$ROOT_DIR/etc/pki/tls/private";
 mkdir -p "$ROOT_DIR/opt/traffic_ops/app/public/routing";
 mkdir "$ROOT_DIR/opt/traffic_ops/app/db";
 mkdir -p "$ROOT_DIR/opt/traffic_ops/app/conf/production";
-echo "{\"hypnotoad\":{\"listen\":[\"https://[::]:60443?cert=$ROOT_DIR/etc/pki/tls/certs/localhost.crt&key=$ROOT_DIR/etc/pki/tls/private/localhost.key\"]}}" > "$ROOT_DIR/opt/traffic_ops/app/conf/cdn.conf";
+cat > "$ROOT_DIR/opt/traffic_ops/app/conf/cdn.conf" <<EOF
+{
+       "hypnotoad": {
+               "listen": [
+                       "https://[::]:60443?cert=$ROOT_DIR/etc/pki/tls/certs/localhost.crt&key=$ROOT_DIR/etc/pki/tls/private/localhost.key"
+               ]
+       }
+}
+EOF;
 mkdir -p "$ROOT_DIR/opt/traffic_ops/install/data/json";
 mkdir "$ROOT_DIR/opt/traffic_ops/install/bin";
```

causes:

```
Error loading users.json file: [Errno 2] No such file or directory: '/tmp/tmp.7vg08w3Ow2/opt/traffic_ops/install/data/json/users.json'
```

I don't know how or why but it does

zrhoffman replied:
If the delimiter ends with a semicolon like `EOF;`, it interprets the semicolon as part of the delimiter.

ocket8888 replied:
I have no idea how that caused that error, but it's fixed now.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+}
+EOF
+
+"$MY_DIR/postinstall.py" --no-root --root-directory="$ROOT_DIR" --no-restart-to --no-database --ops-user="$(whoami)" --ops-group="$(whoami)" --automatic --cfile="$ROOT_DIR/defaults.json" --debug 2>"$ROOT_DIR/stderr" | tee "$ROOT_DIR/stdout"
```
`whoami` for the user and group can be `id -un` and `id -gn`, respectively

ocket8888 replied:
I'll replace the group one, but the `whoami` manpage says

> Same as id -un.

So there's no reason to replace that one

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+ROOT_DIR="$(mktemp -d)";
```
`ROOT_DIR` should not be all caps because it is not an environment variable.

ocket8888 replied:
says who?

zrhoffman replied:
From [Google's shell style guide](https://github.com/google/styleguide/blob/gh-pages/shellguide.md#constants-and-environment-variable-names):

>### Constants and Environment Variable Names
>
>All caps, separated with underscores, declared at the top of the file.
>
>Constants and anything exported to the environment should be
>capitalized.
>
>```shell
># Constant
>readonly PATH_TO_FILES='/some/path'
>
># Both constant and environment
>declare -xr ORACLE_SID='PROD'
>```

In other words, if you want ROOT_DIR to be a constant, it should
* Be declared using `readonly` or `declare -r`
* Be declared at the top of the file instead of inline

zrhoffman replied:
Granted, declaring inline vs at the top is more subjective.

ocket8888 replied:
Google ain't the boss of me. But I don't have a problem marking things that should be readonly with `readonly`.

ocket8888 replied:
I resolved all the other conversations about the same thing so we can just discuss it here

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+mkdir -p "$ROOT_DIR/opt/traffic_ops/install/data/json";
+mkdir "$ROOT_DIR/opt/traffic_ops/install/bin";
+
+MY_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
`MY_DIR` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+
+/usr/bin/python3 -c "$USERS_TEST";
+
+POST_INSTALL_JSON="$ROOT_DIR/opt/traffic_ops/install/data/json/post_install.json";
```
`POST_INSTALL_JSON` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+PROFILES_JSON_EXPECTED="{
```
`PROFILES_JSON_EXPECTED` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	\"dns_subdomain\": \"cdn1.kabletown.net\"
+}";
+
+PROFILES_JSON_ACTUAL="$(cat $ROOT_DIR/opt/traffic_ops/install/data/json/profiles.json)";
```
`PROFILES_JSON_ACTUAL` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+DB_CONF_EXPECTED="production:
```
`DB_CONF_EXPECTED` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+    driver: postgres
+    open: host=localhost port=5432 user=traffic_ops password=twelve dbname=traffic_ops sslmode=disable";
+
+DB_CONF_ACTUAL="$(cat $ROOT_DIR/opt/traffic_ops/app/db/dbconf.yml)";
```
`DB_CONF_ACTUAL` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+CDN_CONF_TEST="
```
`CDN_CON_TEST` should not be all caps because it is not an environment variable.

zrhoffman replied:
Instead of storing this script to `CDN_CONF_TEST` and using it as a really long argument to `python -c`, if this script is a here document, `/usr/bin/python3` can be the command that the here document is standard input for.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+/usr/bin/python3 -c "#CDN_CONF_TEST";
+
+
+DATABASE_CONF_EXPECTED='{
```
`DATABASE_CONF_EXPECTED` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	"description": "Pg database on localhost:5432"
+}';
+
+DATABASE_CONF_ACTUAL="$(cat $ROOT_DIR/opt/traffic_ops/app/conf/production/database.conf)";
```
`DATABASE_CONF_ACTUAL` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+CSR_FILE="$ROOT_DIR/etc/pki/tls/certs/localhost.csr";
```
`CSR_FILE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+fi
+
+CSR_FILE="$ROOT_DIR/etc/pki/tls/certs/localhost.csr";
+CSR_FILE_TYPE="$(file $CSR_FILE)";
```
`CSR_FILE_TYPE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+CERT_FILE="$ROOT_DIR/etc/pki/tls/certs/localhost.crt";
```
`CERT_FILE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+fi
+
+CERT_FILE="$ROOT_DIR/etc/pki/tls/certs/localhost.crt";
+CERT_FILE_TYPE="$(file $CERT_FILE)";
```
`CERT_FILE_TYPE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+	exit 1;
+fi
+
+KEY_FILE="$ROOT_DIR/etc/pki/tls/private/localhost.key";
```
`KEY_FILE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+fi
+
+KEY_FILE="$ROOT_DIR/etc/pki/tls/private/localhost.key";
+KEY_FILE_TYPE="$(file $KEY_FILE)";
```
`KEY_FILE_TYPE` should not be all caps because it is not an environment variable.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+fi
+
+USERS_JSON_FILE="$ROOT_DIR/opt/traffic_ops/install/data/json/users.json";
+USERS_TEST="
+import json
+import sys
+
+try:
+	with open('$USERS_JSON_FILE') as fd:
+		users_json = json.load(fd)
+except Exception as e:
+	print('Error loading users.json file:', e, file=sys.stderr)
+	exit(1)
+
+if not isinstance(users_json, dict) or len(users_json) != 2 or 'username' not in users_json or 'password' not in users_json:
+	print('Malformed users.json file - not an object or incorrect keys', file=sys.stderr)
+	exit(1)
+
+username = users_json['username']
+if not isinstance(username, str):
+	print('Username is not a string in users.json:', username, file=sys.stderr)
+	exit(1)
+
+if username != 'admin':
+	print('Incorrect username in users.json, expected: admin, got:', username, file=sys.stderr)
+	exit(1)
+
+password = users_json['password']
+if not isinstance(password, str):
+	print('Password is not a string in users.json:', password, file=sys.stderr)
+	exit(1)
+
+if not password.startswith('SCRYPT:16384:8:1:') or len(password.split(':')) != 6:
+	print('Malformed password field in users.json:', password, file=sys.stderr)
+	exit(1)
+
+exit(0)
+"
+
+/usr/bin/python3 -c "$USERS_TEST";
```
Instead of storing this script to `USERS_TEST` and using it as a really long argument to `python -c`, if this script is a here document, `/usr/bin/python3` can be the command that the here document is standard input for.

zrhoffman commented on traffic_ops/install/bin/postinstall.test.sh:
```diff
@@ -0,0 +1,424 @@
+
+MY_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+cat <<- EOF > "$ROOT_DIR/defaults.json"
```
`EOF` is not a very descriptive delimiter, is there a delimiter that better describes the here document's contents?

ocket8888 replied:
Does it need to be descriptive? It is replacing an end-of-file character. To be honest, I find delimiters *besides* EOF confusing.

zrhoffman replied:
The description doesn't need to be in the delimiter. But if it's not, there should be a comment that explains the significance of the content redirected to defaults.json.

ocket8888 replied:
A comment sounds like a much better place for descriptions of things.

ocket8888 commented:
> Python 2.7 is supported by Red Hat through June 2024. Although we definitely don't need to support Python 2.7 just for this script to exist in the repo, there would be a much stronger case for this Python script to replace the Perl one if it could be run using both Python 2.7 and 3.6.

But ATC as a project doesn't support Python2.7 as of 4.0.0

zrhoffman commented:
> But ATC as a project doesn't support Python2.7 as of 4.0.0

Not trying to convince you, just describing the conditions under which it
could replace the Perl postinstall.

On Tue, Nov 10, 2020 at 12:42 PM ocket8888 <notifications@github.com> wrote:

> Python 2.7 is supported by Red Hat through June 2024. Although we
> definitely don't need to support Python 2.7 just for this script to exist
> in the repo, there would be a much stronger case for this Python script to
> replace the Perl one if it could be run using both Python 2.7 and 3.6.
>
> But ATC as a project doesn't support Python2.7 as of 4.0.0
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/trafficcontrol/pull/4763#issuecomment-724924255>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACVFRL2UTPLRQYKUSYTHNLTSPGJRPANCNFSM4NX4BGAA>
> .
>

zrhoffman commented:
Everything looks good so far, just need to test

zrhoffman reviewed (changes requested):
I'm getting an error on CentOS 8.2 using Python 3.6.8:
```
[root@trafficops-perl app]# /usr/libexec/platform-python ../install/bin/postinstall.py -a --cfile postinstall.json
INFO:root:Starting postinstall
INFO:root:Debug is on
INFO:root:Running in automatic mode
INFO:root:Using input file postinstall.json
INFO:root:Adding question 'Port to serve on?' with default answer 443
INFO:root:Adding question 'ldap.conf location?' with default answer /opt/traffic_ops/app/conf/ldap.conf
INFO:root:Adding question 'LDAP server hostname' with default answer
INFO:root:Adding question 'LDAP Admin DN' with default answer
INFO:root:Adding question 'LDAP Admin Password' with default answer
INFO:root:Adding question 'LDAP Search Base' with default answer
INFO:root:Adding question 'LDAP Search Query' with default answer
INFO:root:Adding question 'LDAP Skip TLS verify' with default answer
INFO:root:Adding question 'LDAP Timeout Seconds' with default answer
INFO:root:Adding question 'Country Name (2 letter code)' with default answer
INFO:root:Adding question 'State or Province Name (full name)' with default answer
INFO:root:Adding question 'Locality Name (eg, city)' with default answer
INFO:root:Adding question 'Organization Name (eg, company)' with default answer
INFO:root:Adding question 'Organizational Unit Name (eg, section)' with default answer
INFO:root:Adding question 'Common Name (eg, your name or your server's hostname)' with default answer
INFO:root:Adding question 'RSA Passphrase' with default answer
INFO:root:File sanity check complete - found 16 differences
INFO:root:===========/opt/traffic_ops/app/conf/production/database.conf===========
INFO:root:Database configuration has been saved
INFO:root:===========/opt/traffic_ops/app/db/dbconf.yml===========
INFO:root:Not setting up ldap
INFO:root:===========/opt/traffic_ops/install/data/json/users.json===========
INFO:root:===========/opt/traffic_ops/install/data/json/openssl_configuration.json===========
INFO:root:===========/opt/traffic_ops/install/data/json/profiles.json===========
INFO:root:Not downloading Maxmind data
INFO:root:Not generating openssl certification
INFO:root:===========/opt/traffic_ops/app/conf/cdn.conf===========
INFO:root:CDN configuration has been saved
INFO:root:paramconf {'tm.url': 'https://trafficops.infra.ciab.test:443', 'cdn_name': 'CDN-in-a-Box', 'dns_subdomain': 'mycdn.ciab.test'}
INFO:root:Setting up the database data
Traceback (most recent call last):
  File "/opt/traffic_ops/install/bin/postinstall.py", line 1241, in <module>
    ARGS.no_database
  File "/opt/traffic_ops/install/bin/postinstall.py", line 1140, in main
    setup_database_data(conn_str, admin_conf, paramconf, root_dir)
  File "/opt/traffic_ops/install/bin/postinstall.py", line 912, in setup_database_data
    if exec_psql(conn_str, tables_found_query) == "t":
  File "/opt/traffic_ops/install/bin/postinstall.py", line 877, in exec_psql
    proc = subprocess.run(cmd, capture_output=True, universal_newlines=True, check=False)
  File "/usr/lib64/python3.6/subprocess.py", line 423, in run
    with Popen(*popenargs, **kwargs) as process:
TypeError: __init__() got an unexpected keyword argument 'capture_output'
```

Using this cfile, which works with the Perl `_postinstall` script, and testing on the [postinstall-python-centos8](/zrhoffman/trafficcontrol/commits/postinstall-python-centos8) branch:
```json
{
  "/opt/traffic_ops/app/conf/cdn.conf": [
    {
      "Generate a new secret?": "no",
      "config_var": "genSecret"
    },
    {
      "Number of secrets to keep?": "1",
      "config_var": "keepSecrets"
    },
    {
      "Number of workers?": "12",
      "config_var": "workers"
    },
    {
      "Traffic Ops url?": "https://trafficops.infra.ciab.test:443",
      "config_var": "base_url"
    }
  ],
  "/opt/traffic_ops/app/conf/ldap.conf": [
    {
      "Do you want to set up LDAP?": "no",
      "config_var": "setupLdap"
    }
  ],
  "/opt/traffic_ops/app/conf/production/database.conf": [
    {
      "Database type": "Pg",
      "config_var": "type"
    },
    {
      "Database name": "traffic_ops",
      "config_var": "dbname"
    },
    {
      "Database server hostname IP or FQDN": "db.infra.ciab.test",
      "config_var": "hostname"
    },
    {
      "Database port number": "5432",
      "config_var": "port"
    },
    {
      "Traffic Ops database user": "traffic_ops",
      "config_var": "user"
    },
    {
      "Traffic Ops database password": "twelve",
      "config_var": "password",
      "hidden": "1"
    }
  ],
  "/opt/traffic_ops/app/db/dbconf.yml": [
    {
      "Database server root (admin) username": "traffic_ops",
      "config_var": "pgUser"
    },
    {
      "Database server admin password": "twelve",
      "config_var": "pgPassword",
      "hidden": "1"
    },
    {
      "Download Maxmind Database?": "no",
      "config_var": "maxmind"
    }
  ],
  "/opt/traffic_ops/install/data/json/openssl_configuration.json": [
    {
      "Do you want to generate a certificate?": "no",
      "config_var": "genCert"
    }
  ],
  "/opt/traffic_ops/install/data/json/profiles.json": [
    {
      "Traffic Ops url": "https://trafficops.infra.ciab.test:443",
      "config_var": "tm.url"
    },
    {
      "Human-readable CDN Name.  (No whitespace, please)": "CDN-in-a-Box",
      "config_var": "cdn_name"
    },
    {
      "DNS sub-domain for which your CDN is authoritative": "mycdn.ciab.test",
      "config_var": "dns_subdomain"
    }
  ],
  "/opt/traffic_ops/install/data/json/users.json": [
    {
      "Administration username for Traffic Ops": "admin",
      "config_var": "tmAdminUser"
    },
    {
      "Password for the admin user": "twelve12",
      "config_var": "tmAdminPw",
      "hidden": "1"
    }
  ],
  "/opt/traffic_ops/install/data/profiles/": [
    {
      "Add custom profiles?": "no",
      "config_var": "custom_profiles"
    }
  ]
}
```

ocket8888 commented on traffic_ops/install/bin/postinstall.py:
```diff
@@ -0,0 +1,1247 @@
+-a, --automatic               If there are questions in the config file which do not have answers,
+                              the script will look to the defaults for the answer. If the answer is
+                              not in the defaults the script will exit.
+--cfile [FILE]                An input config file used to ask and answer questions.
+--debug                       Enables verbose logging output.
+--defaults [FILE]             Writes out a configuration file with defaults which can be used as
```
I actually think POSIX is fine with that, it's GNU non-compliant. At any rate I'm not sure `argparse` can do that, because it parses strings of characters afters a `-` as separate short-form options as per GNU spec. I'll need to look into that.

ocket8888 replied:
[Looks like argparse does, in fact, support this](https://stackoverflow.com/questions/31127366/single-dash-for-argparse-long-options)

ocket8888 commented:
I think that's because the Perl script will silently fail to apply database configuration if no database is present, which is why the Python script now comes with a command-line option to skip database operations. Silent failures are a bug, IMO.

zrhoffman commented:
> I think that's because the Perl script will silently fail to apply database configuration if no database is present, which is why the Python script now comes with a command-line option to skip database operations. Silent failures are a bug, IMO.

But there *is* a database present, and the Perl script does succeed. Pushed 337785b0f0 to the [postinstall-python-centos8](/zrhoffman/trafficcontrol/commits/postinstall-python-centos8) branch, so now `trafficops-perl` initializes the database using the Perl `_postinstall` so you can observe it succeeding.

ocket8888 commented:
Looks like the error stems from my use of `capture_output`, which was added in Python 3.7. Currently the oldest minor release still receiving security fixes is 3.6 - and that's what's in the CentOS8 `/usr/libexec/` - so we should definitely still support it. Fixed.

Also added support for a Python interpreter at `/usr/libexec/platform-python` - but not Python 2 support. Python 2 is dead and I'm no necromancer. If it would help adoption I would not be in any way opposed to making the script support Python 2 until CentOS 7's end-of-life either after this is merged or through a PR directly to my branch before it's merged, but it's work I don't want to do. I like my type hints, and part of the point of doing this rewrite was to eliminate dependencies outside of the standard library which supporting the two cleanly might require (like `future` or `six`) .

zrhoffman commented on traffic_ops/install/bin/postinstall.py:
```diff
@@ -0,0 +1,1344 @@
+#!/usr/bin/env -S sh -c 'PATH="${PATH}:/usr/libexec" exec "$(type -p python platform-python | head -n1)" "$0" "$@"'
```
Testing it out, ^ needs a semicolon after the PATH addition in order to work, like

```shell
#!/usr/bin/env -S sh -c 'PATH="${PATH}:/usr/libexec"; exec "$(type -p python3 platform-python | head -n1)" "$0" "$@"'
```

However, apparently `env -S` is not supported on the GNU `env` that CentOS 7 uses (8.22), and since CentOS 8 goes EOL 2.5 years before CentOS 7 does, we should support CentOS 7's `env`. This would work:

```bash
#!/usr/bin/env bash
"exec" "bash" "-c" "PATH+=:/usr/libexec/ exec \$(type -p python3 platform-python | head -n1) \"$0\" $*"
```

ocket8888 replied:
Turning a Python script into a polyglot bash script seems convoluted. Can I not just use `#!/usr/bin/env python3` or `#!/bin/sh -c 'PATH="${PATH}:/usr/libexec"; exec "$(type -p python3 platform-python | head -n1)" "$0" "$@"'` ? Why does `env` need so much hand-holding? If `platform-python` isn't in the `PATH` should I be using it at all? [The RedHat Developer Blog](https://developers.redhat.com/blog/2018/11/14/python-in-rhel-8/) says no:

> _"If you are not re-building the distro, do not use Platform-Python directly. Install `python3` and use that."_

and later reiterates:

> _"Do not use platform-python for your applications. However, use platform-python if you are writing system/admin code for RHEL 8."_

zrhoffman replied:
> Turning a Python script into a polyglot bash script seems convoluted. Can I not just use `#!/usr/bin/env python3` or `#!/bin/sh -c 'PATH="${PATH}:/usr/libexec"; exec "$(type -p python3 platform-python | head -n1)" "$0" "$@"'` ? Why does `env` need so much hand-holding?

It's not `env` that needs the handholding, it's shebang. Shebang does not split arguments by a *space* character (or by anything else), so in the snippet you posted above, `-c 'PATH="${PATH}:/usr/libexec"; exec "$(type -p python3 platform-python | head -n1)" "$0" "$@"'` is all treated as a single argument, which results in an error:

```shell
[root@732dff90b85f a]# ./a.py
/bin/sh: - : invalid option
Usage:  /bin/sh [GNU long option] [option] ...
        /bin/sh [GNU long option] [option] script-file ...
GNU long options:
        --debug
        --debugger
        --dump-po-strings
        --dump-strings
        --help
        --init-file
        --login
        --noediting
        --noprofile
        --norc
        --posix
        --rcfile
        --rpm-requires
        --restricted
        --verbose
        --version
Shell options:
        -ilrsD or -c command or -O shopt_option         (invocation only)
        -abefhkmnptuvxBCHP or -o option
```


> If `platform-python` isn't in the `PATH` should I be using it at all? [The RedHat Developer Blog](https://developers.redhat.com/blog/2018/11/14/python-in-rhel-8/) says no:
> 
> > _"If you are not re-building the distro, do not use Platform-Python directly. Install `python3` and use that."_
> 
> and later reiterates:
> 
> > _"Do not use platform-python for your applications. However, use platform-python if you are writing system/admin code for RHEL 8."_

Since there is no Python 2 support currently, yeah, might as well just use `#!/usr/bin/env python3` directly for now.

ocket8888 replied:
> _"Since there is no Python 2 support currently, yeah, might as well just use `#!/usr/bin/env python3` directly for now."_

Then that's what I'll do

zrhoffman commented on traffic_ops/install/bin/postinstall.py:
```diff
@@ -0,0 +1,1344 @@
+#!/usr/bin/env -S sh -c 'PATH="${PATH}:/usr/libexec" exec "$(type -p python platform-python | head -n1)" "$0" "$@"'
```
The Python postinstall script does not support Python 2.7, which is fine, but because it doesn't, it should search the patch for a `python3` explicitly, not `python`, else it would not work on CentOS 7.

ocket8888 replied:
Done

zrhoffman reviewed (approved):
Looks great
